quint/src/runtime/trace.ts

```diff
@@ -140,153 +140,200 @@ export const noExecutionListener: ExecutionListener = {
   onRunReturn: (_outcome: Maybe<EvalResult>, _trace: EvalResult[]) => {},
 }
 
+/**
+ * The interface of a trace recorder that records the best trace.
+ */
 export interface TraceRecorder extends ExecutionListener {
+  /**
+   * The verbosity level used by the recorder (mutable).
+   */
+  verbosityLevel: number
+  /**
+   * The random number generator used by the recorder (mutable).
+   */
+  rng: Rng
+
+  /**
+   * Clear the recorded trace
+   */
+  clear: () => void
+
+  /**
+   * Get the best recorded trace.
+   * @returns the best recorded trace.
+   */
   getBestTrace: () => ExecutionFrame
+
+  /**
+   * Get the seed of the best recorded trace.
+   * @returns the seed that was used to generate the best trace.
+   */
   getBestTraceSeed: () => bigint
 }
 
 // a trace recording listener
-export const newTraceRecorder = (verbosityLevel: number, rng: Rng) => {
-  // the bottom frame encodes the whole trace
-  const bottomFrame = (): ExecutionFrame => {
-    return {
-      // this is just a dummy operator application
-      app: { id: 0n, kind: 'app', opcode: 'Rec', args: [] },
-      // we will store the sequence of states here
-      args: [],
-      // the result of the trace evaluation
-      result: just(rv.mkBool(true)),
-      // and here we store the subframes for the top-level actions
-      subframes: [],
-    }
-  }
+export const newTraceRecorder = (verbosityLevel: number, rng: Rng): TraceRecorder => {
+  return new TraceRecorderImpl(verbosityLevel, rng)
+}
 
+// a private implementation of a trace recorder
+class TraceRecorderImpl implements TraceRecorder {
+  verbosityLevel: number
+  rng: Rng
   // the best trace is stored here
-  let bestTrace = bottomFrame()
+  private bestTrace: ExecutionFrame
   // the seed value for the best trace is stored here
-  let bestTraceSeed = rng.getState()
+  private bestTraceSeed: bigint
   // whenever a run is entered, we store its seed here
-  let runSeed = bestTraceSeed
+  private runSeed: bigint
   // during simulation, a trace is built here
-  let frameStack: ExecutionFrame[] = [bestTrace]
-
-  return {
-    getBestTrace: (): ExecutionFrame => {
-      return bestTrace
```

Comment:
Most of the changes below are just about moving the code in the class definition